vllm/v1/worker/gpu_model_runner.py

```diff
@@ -820,6 +822,28 @@ def make_sampling_metadata(
                 self.top_p_cpu_tensor[:self.num_reqs], non_blocking=True)
             self.top_k[:self.num_reqs].copy_(
                 self.top_k_cpu_tensor[:self.num_reqs], non_blocking=True)
+
+        output_token_ids: List[List[int]] = []
+        prompt_token_ids: List[List[int]] = []
+        frequency_penalties: List[float] = []
+        presence_penalties: List[float] = []
+        repetition_penalties: List[float] = []
+        min_tokens: List[int] = []
+        stop_token_ids: List[set[int]] = []
```

Comment:
QQ: Do we have to use this python representations? Now that the worker is stateful with the "persistent batch", can we keep these in tensors?

Reply:
**Tensor representation**
I think we can convert frequency_penalties, presence_penalties, repetition_penalties, prompt_token_ids into tensors because they won't change in every step. 

**Non Tensor representation**
- output_token_ids will be updated in every step and so we might need to maintain it as a list. 
- min_tokens and stop_token_ids are not used directly. They are used in combination with output_token_ids to determine the the indices of the logits that need to be set to -inf.

Do you think it would be ok to convert   frequency_penalties, presence_penalties, repetition_penalties, prompt_token_ids  into tensor representation and retain the rest (output_token_ids, min_tokens, stop_token_ids) as python representations?

Reply:
@sroy745 Thanks for the explanation. Sounds good to me!

Reply:
In the future, we could do more optimizations based on the states. For example, we can actually cache and update the statistics (histogram), so that we don't need to recompute it every step.

Reply:
@WoosukKwon as discussed moved frequency_penalties, presence_penalties, repetition_penalties, prompt_token_ids into tensor representations. I will address the incremental update of output_token statistics in a separate pr (I have added a TODO for it). PTAL.

I am working on adding some tests for sampling metadata management in InputBatch.